hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/analysis/HoodieAnalysis.scala

```diff
@@ -181,6 +183,67 @@ object HoodieAnalysis extends SparkAdapterSupport {
    */
   case class AdaptIngestionTargetLogicalRelations(spark: SparkSession) extends Rule[LogicalPlan] {
 
+    /**
+     * The function updates the HadoopFSRelation so that it uses HoodieFileIndex instead of
+     * HoodieFileIndexTimestampKeyGen. Also the data type for output attributes of the plan are changed
+     * accordingly. HoodieFileIndexTimestampKeyGen is used by HoodieBaseRelation for reading tables with
+     * Timestamp or custom key generator.
+     */
+    private def transformReaderFSRelation(logicalPlan: Option[LogicalPlan]): Option[LogicalPlan] = {
+      def getAttributesFromTableSchema(catalogTableOpt: Option[CatalogTable], lr: LogicalRelation, attributesSet: Set[AttributeReference]) = {
+        var finalAttrs: List[AttributeReference] = List.empty
+        if (catalogTableOpt.isDefined) {
+          for (attr <- lr.output) {
+            val origAttr: AttributeReference = attributesSet.collectFirst({ case a if a.name.equals(attr.name) => a }).get
+            val catalogAttr = catalogTableOpt.get.partitionSchema.fields.collectFirst({ case a if a.name.equals(attr.name) => a })
+            val newAttr: AttributeReference = if (catalogAttr.isDefined) {
+              origAttr.copy(dataType = catalogAttr.get.dataType)(origAttr.exprId, origAttr.qualifier)
+            } else {
+              origAttr
+            }
+            finalAttrs = finalAttrs :+ newAttr
+          }
+        }
+        finalAttrs
+      }
+
+      def getHadoopFsRelation(plan: LogicalPlan): Option[HadoopFsRelation] = {
+        EliminateSubqueryAliases(plan) match {
+          // First, we need to weed out unresolved plans
+          case plan if !plan.resolved => None
+          // NOTE: When resolving Hudi table we allow [[Filter]]s and [[Project]]s be applied
+          //       on top of it
+          case PhysicalOperation(_, _, LogicalRelation(relation, _, _, _)) if relation.isInstanceOf[HadoopFsRelation] => Some(relation.asInstanceOf[HadoopFsRelation])
+          case _ => None
+        }
+      }
+
+      logicalPlan.map(relation => {
+        val catalogTableOpt = sparkAdapter.resolveHoodieTable(relation)
+        val fsRelation = getHadoopFsRelation(relation)
+        if (fsRelation.isDefined && fsRelation.get.location.isInstanceOf[HoodieReaderFileIndex]) {
```

Comment:
under what scenarios, index is an instance of `HoodieReaderFileIndex`? Is it only for insert into, update and merge into commands? Let's add in a comment.

Reply:
I checked the other commands and it is also required for DeleteFromTable and CreateTable. Included all commands which take in a logical plan as argument